tests/test_models.py

```diff
@@ -0,0 +1,237 @@
+from django.test import TestCase
+
+import pytest
+
+from user_tasks.exceptions import TaskCanceledException
+from user_tasks.models import UserTaskArtifact, UserTaskStatus
+
+LOGGER = logging.getLogger(__name__)
+
+
+class TestUserTaskStatus(TestCase):
+    """Tests of the UserTask model."""
+
+    @classmethod
+    def setUpTestData(cls):
+        cls.user = User.objects.create_user('test_user', 'test@example.com', 'password')
+
+    def test_cancel_parent(self):
+        """Canceling a container status should also cancel its children."""
+        parent = self._status(is_container=True, state=UserTaskStatus.PENDING)
+        status = self._status(parent=parent, state=UserTaskStatus.PENDING)
+        parent.cancel()
+        status.refresh_from_db()
+        assert parent.state == UserTaskStatus.CANCELED
+        assert status.state == UserTaskStatus.CANCELED
+
+    def test_cancel_finished_task(self):
+        """Attempting to cancel an already-finished task should have no effect."""
+        status = self._status(state=UserTaskStatus.SUCCEEDED)
+        status.cancel()
+        assert status.state == UserTaskStatus.SUCCEEDED
+
+    def test_fail_with_parent(self):
+        """A task which fails should also mark any parent as having failed."""
+        parent = self._status(is_container=True)
+        status = self._status(parent=parent)
+        status.fail('Oops!')
+        parent.refresh_from_db()
+        assert status.state == UserTaskStatus.FAILED
+        assert parent.state == UserTaskStatus.FAILED
+        artifact = UserTaskArtifact.objects.get(status=status)
+        assert artifact.text == 'Oops!'
+        parent_artifact = UserTaskArtifact.objects.get(status=parent)
+        assert parent_artifact.text == 'Oops!'
+
+    def test_increment_completed_steps(self):
+        """increment_completed_steps() should update both the current status and any parent status."""
+        parent = self._status(is_container=True, total_steps=0)
+        status = self._status(parent=parent, total_steps=2)
+        status.increment_completed_steps()
+        parent.refresh_from_db()
+        assert status.completed_steps == 1
+        assert parent.completed_steps == 1
+
+    def test_increment_completed_steps_explicit_steps(self):
+        """increment_completed_steps() should support explicit step counts greater than 1."""
+        status = self._status(total_steps=5)
+        status.increment_completed_steps(3)
+        assert status.completed_steps == 3
+        status.refresh_from_db()
+        assert status.completed_steps == 3
+
+    def test_increment_total_steps(self):
+        """increment_total_steps() should correctly update the status and any parent."""
+        parent = self._status(is_container=True, total_steps=3)
+        status = self._status(parent=parent, total_steps=1)
+        status.increment_total_steps(2)
+        parent.refresh_from_db()
+        assert status.total_steps == 3
+        assert parent.total_steps == 5
```

Comment:
I omitted any hypothetical child tasks not directly relevant to the test case, but I suppose that's confusing.  I'll add another child with total_steps=2 for clarity.